src/coreclr/jit/llvm.cpp

```diff
@@ -2648,14 +2730,60 @@ void Llvm::lowerStoreLcl(GenTreeLclVarCommon* storeLclNode)
     }
 }
 
+void Llvm::lowerFieldOfDependentlyPromotedStruct(GenTree* node)
+{
+    if (node->OperIs(GT_STORE_LCL_VAR, GT_LCL_VAR, GT_LCL_VAR_ADDR))
+    {
+        // fail unsupported promoted structs
+        GenTreeLclVarCommon* lclVar = node->AsLclVarCommon();
+        LclVarDsc*           varDsc = _compiler->lvaGetDesc(lclVar->GetLclNum());
+
+        if (_compiler->lvaIsFieldOfDependentlyPromotedStruct(varDsc))
+        {
+            switch (node->OperGet())
+            {
+                case GT_LCL_VAR:
+                    lclVar->SetOper(GT_LCL_FLD);
+                    break;
+
+                case GT_STORE_LCL_VAR:
+                    lclVar->SetOper(GT_STORE_LCL_FLD);
+                    break;
+#
+                case GT_LCL_VAR_ADDR:
+                    lclVar->SetOper(GT_LCL_FLD_ADDR);
+                    break;
+            }
+
+            lclVar->AsLclFld()->SetLclOffs(varDsc->lvFldOffset);
+            lclVar->SetLclNum(varDsc->lvParentLcl);
+
+            if (node->OperIs(GT_STORE_LCL_FLD, GT_LCL_FLD_ADDR) &&
+                node->IsPartialLclFld(_compiler) &&
+                node->gtFlags & GTF_VAR_DEF)
+            {
+                node->gtFlags |= GTF_VAR_USEASG;
+            }
```

Comment:
```suggestion
    if (node->OperIsLocal() || node->OperIsLocalAddr())
    {
        GenTreeLclVarCommon* lclVar = node->AsLclVarCommon();
        uint16_t             offset = lclVar->GetLclOffs();
        LclVarDsc*           varDsc = _compiler->lvaGetDesc(lclVar->GetLclNum());        

        if (_compiler->lvaIsFieldOfDependentlyPromotedStruct(varDsc))
        {
            switch (node->OperGet())
            {
                case GT_LCL_VAR:
                    lclVar->SetOper(GT_LCL_FLD);
                    break;

                case GT_STORE_LCL_VAR:
                    lclVar->SetOper(GT_STORE_LCL_FLD);
                    break;

                case GT_LCL_VAR_ADDR:
                    lclVar->SetOper(GT_LCL_FLD_ADDR);
                    break;
            }

            lclVar->SetLclNum(varDsc->lvParentLcl);
            lclVar->AsLclFld()->SetLclOffs(varDsc->lvFldOffset + offset );

            if ((node->gtFlags & GTF_VAR_DEF) != 0)
            {
                // Conservatively assume these become partial.
                // TODO-ADDR: only apply to stores be precise.
                node->gtFlags |= GTF_VAR_USEASG;
            }
```
Turns out we cannot really be precise about `GTF_VAR_USEASG` at the moment without significantly more code (to look up the parent store).